src/json/next/serialization.cr

```diff
@@ -0,0 +1,217 @@
+module JSON
+  annotation Field
+  end
+
+  module Serializable
+    annotation Options
+    end
+
+    macro included
+      def self.new(pull : ::JSON::PullParser)
+        instance = allocate
+        instance.initialize(pull)
+        GC.add_finalizer(instance) if instance.responds_to?(:finalize)
+        instance
+      end
+
+      macro inherited
+        include JSON::Serializable
```

Comment:
ditto

Reply:
So when you subclass a type that includes the module, it will also include the module and then define an `initialize` method that can be called. Method lookup for `new` and `initialize` don't look up in superclasses. But I'm starting to think that's probably wrong, or maybe wrong. But I don't know how to fix it. That rules makes it super hard to make sure instance vars are correctly initialized in types. Maybe constructors should be inherited and they should stop working if not all instance vars are initialized. I don't know.

Reply:
So there's a remaining unfixed bug with `{% @type }%` in constructors after #6077? I'm fine with merging hacks if we understand them...

Reply:
Yes, #6077 fixed something different. Isn't it obvious by looking at the code? :-P (just kidding)

In this PR, I honestly don't understand why the hacks work.